k8s: init.sh fails with `Unable to connect to the server: net/http: TLS handshake timeout`

I've seen CI jobs failing because the `integration/kubernetest/init.sh` could not bring up the environment correctly. The error message on the log is always "_Unable to connect to the server: net/http: TLS handshake timeout_".

For example, the [build #1092](http://jenkins.katacontainers.io/job/kata-containers-2.0-tests-fedora-PR-crio-k8s-e2e-minimal/1092) of _kata-containers-2.0-tests-fedora-PR-crio-k8s-e2e-minimal_ , failed with that error:
```
17:36:49 [init.sh:285] INFO: Clean up any leftover CNI configuration
17:36:49 Device "cni0" does not exist.
17:36:49 [init.sh:288] INFO: Start crio service
17:36:49 Created symlink /etc/systemd/system/cri-o.service → /usr/local/lib/systemd/system/crio.service.
17:36:49 Created symlink /etc/systemd/system/multi-user.target.wants/crio.service → /usr/local/lib/systemd/system/crio.service.
17:36:54 ● crio.service - Container Runtime Interface for OCI (CRI-O)
17:36:54      Loaded: loaded (/usr/local/lib/systemd/system/crio.service; enabled; vendor preset: disabled)
17:36:54      Active: active (running) since Tue 2021-12-14 20:36:49 UTC; 5s ago
17:36:54        Docs: https://github.com/cri-o/cri-o
17:36:54    Main PID: 111602 (crio)
17:36:54       Tasks: 10
17:36:54      Memory: 49.4M
17:36:54      CGroup: /system.slice/crio.service
17:36:54              └─111602 /usr/local/bin/crio
17:36:54 
17:36:54 Dec 14 20:36:49 fedoradocker32-443f80 crio[111602]: time="2021-12-14 20:36:49.737824781Z" level=info msg="RDT not available in the host system" file="rdt/rdt.go:57"
17:36:54 Dec 14 20:36:49 fedoradocker32-443f80 crio[111602]: time="2021-12-14 20:36:49.738618951Z" level=info msg="Updated default CNI network name to " file="ocicni/ocicni.go:371"
17:36:54 Dec 14 20:36:49 fedoradocker32-443f80 crio[111602]: time="2021-12-14 20:36:49.739741349Z" level=debug msg="reading hooks from /usr/share/containers/oci/hooks.d" file="hooks/read.go:65"
17:36:54 Dec 14 20:36:49 fedoradocker32-443f80 crio[111602]: time="2021-12-14 20:36:49.772406414Z" level=debug msg="Golang's threads limit set to 114930" file="server/server.go:348"
17:36:54 Dec 14 20:36:49 fedoradocker32-443f80 crio[111602]: time="2021-12-14 20:36:49.773019767Z" level=warning msg="Error encountered when checking whether cri-o should wipe images: version file /var/lib/crio/version not found: open /var/lib/crio/version: no such file or directory" file="server/server.go:524"
17:36:54 Dec 14 20:36:49 fedoradocker32-443f80 crio[111602]: time="2021-12-14 20:36:49.773274190Z" level=debug msg="Sandboxes: []" file="server/server.go:496"
17:36:54 Dec 14 20:36:49 fedoradocker32-443f80 crio[111602]: time="2021-12-14 20:36:49.773390700Z" level=debug msg="Registered SIGHUP watcher for config" file="config/reload.go:32"
17:36:54 Dec 14 20:36:49 fedoradocker32-443f80 crio[111602]: time="2021-12-14 20:36:49.773433904Z" level=debug msg="Metrics are disabled" file="server/server.go:507"
17:36:54 Dec 14 20:36:49 fedoradocker32-443f80 crio[111602]: time="2021-12-14 20:36:49.776377262Z" level=debug msg="monitoring \"/usr/share/containers/oci/hooks.d\" for hooks" file="hooks/monitor.go:43"
17:36:54 Dec 14 20:36:49 fedoradocker32-443f80 systemd[1]: Started Container Runtime Interface for OCI (CRI-O).
17:36:54 [init.sh:291] INFO: Start Kubernetes
17:36:54 Init cluster using /var/run/crio/crio.sock
17:36:55 [init] Using Kubernetes version: v1.22.0
17:36:55 [preflight] Running pre-flight checks
17:36:55 [preflight] Pulling images required for setting up a Kubernetes cluster
17:36:55 [preflight] This might take a minute or two, depending on the speed of your internet connection
17:36:55 [preflight] You can also perform this action in beforehand using 'kubeadm config images pull'
17:37:21 [certs] Using certificateDir folder "/etc/kubernetes/pki"
17:37:22 [certs] Generating "ca" certificate and key
17:37:22 [certs] Generating "apiserver" certificate and key
17:37:22 [certs] apiserver serving cert is signed for DNS names [fedoradocker32-443f80 kubernetes kubernetes.default kubernetes.default.svc kubernetes.default.svc.cluster.local] and IPs [10.96.0.1 10.0.0.11]
17:37:22 [certs] Generating "apiserver-kubelet-client" certificate and key
17:37:22 [certs] Generating "front-proxy-ca" certificate and key
17:37:22 [certs] Generating "front-proxy-client" certificate and key
17:37:23 [certs] Generating "etcd/ca" certificate and key
17:37:23 [certs] Generating "etcd/server" certificate and key
17:37:23 [certs] etcd/server serving cert is signed for DNS names [fedoradocker32-443f80 localhost] and IPs [10.0.0.11 127.0.0.1 ::1]
17:37:23 [certs] Generating "etcd/peer" certificate and key
17:37:23 [certs] etcd/peer serving cert is signed for DNS names [fedoradocker32-443f80 localhost] and IPs [10.0.0.11 127.0.0.1 ::1]
17:37:23 [certs] Generating "etcd/healthcheck-client" certificate and key
17:37:24 [certs] Generating "apiserver-etcd-client" certificate and key
17:37:24 [certs] Generating "sa" key and public key
17:37:24 [kubeconfig] Using kubeconfig folder "/etc/kubernetes"
17:37:24 [kubeconfig] Writing "admin.conf" kubeconfig file
17:37:24 [kubeconfig] Writing "kubelet.conf" kubeconfig file
17:37:25 [kubeconfig] Writing "controller-manager.conf" kubeconfig file
17:37:25 [kubeconfig] Writing "scheduler.conf" kubeconfig file
17:37:25 [kubelet-start] Writing kubelet environment file with flags to file "/var/lib/kubelet/kubeadm-flags.env"
17:37:25 [kubelet-start] Writing kubelet configuration to file "/var/lib/kubelet/config.yaml"
17:37:25 [kubelet-start] Starting the kubelet
17:37:25 [control-plane] Using manifest folder "/etc/kubernetes/manifests"
17:37:25 [control-plane] Creating static Pod manifest for "kube-apiserver"
17:37:25 [control-plane] Creating static Pod manifest for "kube-controller-manager"
17:37:25 [control-plane] Creating static Pod manifest for "kube-scheduler"
17:37:25 [etcd] Creating static Pod manifest for local etcd in "/etc/kubernetes/manifests"
17:37:25 [wait-control-plane] Waiting for the kubelet to boot up the control plane as static Pods from directory "/etc/kubernetes/manifests". This can take up to 4m0s
17:37:34 [apiclient] All control plane components are healthy after 8.505587 seconds
17:37:34 [upload-config] Storing the configuration used in ConfigMap "kubeadm-config" in the "kube-system" Namespace
17:37:34 [kubelet] Creating a ConfigMap "kubelet-config-1.22" in namespace kube-system with the configuration for the kubelets in the cluster
17:37:34 [upload-certs] Skipping phase. Please see --upload-certs
17:37:34 [mark-control-plane] Marking the node fedoradocker32-443f80 as control-plane by adding the labels: [node-role.kubernetes.io/master(deprecated) node-role.kubernetes.io/control-plane node.kubernetes.io/exclude-from-external-load-balancers]
17:37:34 [mark-control-plane] Marking the node fedoradocker32-443f80 as control-plane by adding the taints [node-role.kubernetes.io/master:NoSchedule]
17:37:35 [bootstrap-token] Using token: 4ivh97.dyiwfthnpbgcmvy9
17:37:35 [bootstrap-token] Configuring bootstrap tokens, cluster-info ConfigMap, RBAC Roles
17:37:35 [bootstrap-token] configured RBAC rules to allow Node Bootstrap tokens to get nodes
17:37:35 [bootstrap-token] configured RBAC rules to allow Node Bootstrap tokens to post CSRs in order for nodes to get long term certificate credentials
17:37:35 [bootstrap-token] configured RBAC rules to allow the csrapprover controller automatically approve CSRs from a Node Bootstrap Token
17:37:35 [bootstrap-token] configured RBAC rules to allow certificate rotation for all node client certificates in the cluster
17:37:35 [bootstrap-token] Creating the "cluster-info" ConfigMap in the "kube-public" namespace
17:37:35 [kubelet-finalize] Updating "/etc/kubernetes/kubelet.conf" to point to a rotatable kubelet client certificate and key
17:37:35 [addons] Applied essential addon: CoreDNS
17:37:35 [addons] Applied essential addon: kube-proxy
17:37:35 
17:37:35 Your Kubernetes control-plane has initialized successfully!
17:37:35 
17:37:35 To start using your cluster, you need to run the following as a regular user:
17:37:35 
17:37:35   mkdir -p $HOME/.kube
17:37:35   sudo cp -i /etc/kubernetes/admin.conf $HOME/.kube/config
17:37:35   sudo chown $(id -u):$(id -g) $HOME/.kube/config
17:37:35 
17:37:35 Alternatively, if you are the root user, you can run:
17:37:35 
17:37:35   export KUBECONFIG=/etc/kubernetes/admin.conf
17:37:35 
17:37:35 You should now deploy a pod network to the cluster.
17:37:35 Run "kubectl apply -f [podnetwork].yaml" with one of the options listed at:
17:37:35   https://kubernetes.io/docs/concepts/cluster-administration/addons/
17:37:35 
17:37:35 Then you can join any number of worker nodes by running the following on each as root:
17:37:35 
17:37:35 kubeadm join 10.0.0.11:6443 --token 4ivh97.dyiwfthnpbgcmvy9 \
17:37:35 	--discovery-token-ca-cert-hash sha256:c750eb498dcbee26afa2f038723e603a22797387af36204b87fd9fffe39d25ad 
17:37:35 Kubelet options:
17:37:35 KUBELET_KUBEADM_ARGS="--allowed-unsafe-sysctls=kernel.msg*,kernel.shm.*,net.* --container-runtime=remote --container-runtime-endpoint=/var/run/crio/crio.sock --pod-infra-container-image=k8s.gcr.io/pause:3.5 --v=4"
17:38:26 Unable to connect to the server: net/http: TLS handshake timeout
17:39:17 Unable to connect to the server: net/http: TLS handshake timeout
17:40:09 Unable to connect to the server: net/http: TLS handshake timeout
17:41:02 Unable to connect to the server: net/http: TLS handshake timeout
17:41:57 Unable to connect to the server: net/http: TLS handshake timeout
17:41:57 [setup.sh:26] INFO: k8s init failed trying again
17:41:57 [cleanup_env.sh:31] INFO: Reset Kubernetes
17:41:57 [reset] Reading configuration from the cluster...
17:41:57 [reset] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -o yaml'
17:42:37 W1214 20:42:37.445144  112530 reset.go:101] [reset] Unable to fetch the kubeadm-config ConfigMap from cluster: failed to get config map: Get "https://10.0.0.11:6443/api/v1/namespaces/kube-system/configmaps/kubeadm-config?timeout=10s": context deadline exceeded
17:42:37 [preflight] Running pre-flight checks
17:42:37 W1214 20:42:37.445347  112530 removeetcdmember.go:80] [reset] No kubeadm config, using etcd pod spec to get data directory
17:42:37 [reset] Stopping the kubelet service
17:42:37 [reset] Unmounting mounted directories in "/var/lib/kubelet"
17:53:17 W1214 20:53:17.723773  112530 cleanupnode.go:84] [reset] Failed to remove containers: [failed to stop running pod 093b1912442eeb07d677736caafcbd751ae29a2318846a16efa7f8252b701a77: output: time="2021-12-14T20:45:17Z" level=fatal msg="stopping the pod sandbox \"093b1912442eeb07d677736caafcbd751ae29a2318846a16efa7f8252b701a77\": rpc error: code = Unknown desc = failed to stop container for pod sandbox 093b1912442eeb07d677736caafcbd751ae29a2318846a16efa7f8252b701a77: failed to get container 'stopped' status k8s_kube-scheduler_kube-scheduler-fedoradocker32-443f80_kube-system_716d240fcca5094e407f95a3942579f6_0: failed to get container stopped status: 30s timeout reached"
17:53:17 , error: exit status 1, failed to stop running pod 07ef786840e6526bc9d6c4c04b97833cd69cd057abc71658f1f6be2babae6de6: output: time="2021-12-14T20:47:57Z" level=fatal msg="stopping the pod sandbox \"07ef786840e6526bc9d6c4c04b97833cd69cd057abc71658f1f6be2babae6de6\": rpc error: code = Unknown desc = failed to stop container for pod sandbox 07ef786840e6526bc9d6c4c04b97833cd69cd057abc71658f1f6be2babae6de6: failed to get container 'stopped' status k8s_kube-controller-manager_kube-controller-manager-fedoradocker32-443f80_kube-system_1ae0c87ade2ef086cfc13d0ea5d4f3f6_0: failed to get container stopped status: 30s timeout reached"
17:53:17 , error: exit status 1, failed to stop running pod 812c17f9c689edca40ce0cfeb427d1a8d2924ea1c5e20aeb82e758da5971fd33: output: time="2021-12-14T20:50:37Z" level=fatal msg="stopping the pod sandbox \"812c17f9c689edca40ce0cfeb427d1a8d2924ea1c5e20aeb82e758da5971fd33\": rpc error: code = Unknown desc = failed to stop container for pod sandbox 812c17f9c689edca40ce0cfeb427d1a8d2924ea1c5e20aeb82e758da5971fd33: failed to get container 'stopped' status k8s_kube-apiserver_kube-apiserver-fedoradocker32-443f80_kube-system_bb4571d906969fb4d212a18ce64f43a6_0: failed to get container stopped status: 30s timeout reached"
17:53:17 , error: exit status 1, failed to stop running pod b7e915fa6cc68759bd4e069c1a77496975963066c8ef6a7d16968b48d24ebaf2: output: time="2021-12-14T20:53:17Z" level=fatal msg="stopping the pod sandbox \"b7e915fa6cc68759bd4e069c1a77496975963066c8ef6a7d16968b48d24ebaf2\": rpc error: code = Unknown desc = failed to stop container for pod sandbox b7e915fa6cc68759bd4e069c1a77496975963066c8ef6a7d16968b48d24ebaf2: failed to get container 'stopped' status k8s_etcd_etcd-fedoradocker32-443f80_kube-system_4122f50f3c93be2f0cb3a4cc9bc7b9bf_0: failed to get container stopped status: 30s timeout reached"
17:53:17 , error: exit status 1]
17:53:17 [reset] Deleting contents of config directories: [/etc/kubernetes/manifests /etc/kubernetes/pki]
17:53:17 [reset] Deleting files: [/etc/kubernetes/admin.conf /etc/kubernetes/kubelet.conf /etc/kubernetes/bootstrap-kubelet.conf /etc/kubernetes/controller-manager.conf /etc/kubernetes/scheduler.conf]
17:53:17 [reset] Deleting contents of stateful directories: [/var/lib/etcd /var/lib/kubelet /var/lib/dockershim /var/run/kubernetes /var/lib/cni]
17:53:17 
17:53:17 The reset process does not clean CNI configuration. To do so, you must remove /etc/cni/net.d
17:53:17 
17:53:17 The reset process does not reset or clean up iptables rules or IPVS tables.
17:53:17 If you wish to reset iptables, you must do so manually by using the "iptables" command.
17:53:17 
17:53:17 If your cluster was setup to utilize IPVS, run ipvsadm --clear (or similar)
17:53:17 to reset your system's IPVS tables.
17:53:17 
17:53:17 The reset process does not clean your kubeconfig files and you must remove them manually.
17:53:17 Please, check the contents of the $HOME/.kube/config file.
17:53:17 [cleanup_env.sh:35] INFO: Teardown the registry server
17:53:17 [cleanup_env.sh:39] INFO: Stop crio service
17:53:17 [cleanup_env.sh:42] INFO: Remove network devices
17:53:17 remove device: cni0
17:53:17 Cannot find device "cni0"
17:53:17 Cannot find device "cni0"
17:53:17 remove device: flannel.1
17:53:17 Cannot find device "flannel.1"
17:53:18 Cannot find device "flannel.1"
17:53:18 [cleanup_env.sh:54] INFO: Check no kata processes are left behind after reseting kubernetes
17:53:18 [cleanup_env.sh:57] INFO: Checks that pods were not left
17:53:18 Not /run/vc/sbs directory found
17:53:28 [init.sh:279] INFO: Check there aren't dangling processes from previous tests
17:53:29 [init.sh:78] INFO: Build custom stress image
17:53:29 /tmp/jenkins/workspace/kata-containers-2.0-tests-fedora-PR-crio-k8s-e2e-minimal/go/src/github.com/kata-containers/tests/integration/kubernetes/runtimeclass_workloads/stress /tmp/jenkins/workspace/kata-containers-2.0-tests-fedora-PR-crio-k8s-e2e-minimal/go/src/github.com/kata-containers/tests/integration/kubernetes
17:53:29 STEP 1: FROM quay.io/libpod/ubuntu
17:53:29 STEP 2: RUN apt-get -y update &&     apt-get -y upgrade &&     apt-get -y install stress
17:53:29 --> Using cache 2e14401e22a354dddc6dc88d255aa5062c7cb4a4f9be5721ce8b5acd23675df1
17:53:29 STEP 3: COMMIT localhost/stress-kata:latest
17:53:29 --> 2e14401e22a
17:53:29 2e14401e22a354dddc6dc88d255aa5062c7cb4a4f9be5721ce8b5acd23675df1
17:53:29 /tmp/jenkins/workspace/kata-containers-2.0-tests-fedora-PR-crio-k8s-e2e-minimal/go/src/github.com/kata-containers/tests/integration/kubernetes
17:53:29 [init.sh:285] INFO: Clean up any leftover CNI configuration
17:53:29 Device "cni0" does not exist.
17:53:29 [init.sh:288] INFO: Start crio service
17:53:35 ● crio.service - Container Runtime Interface for OCI (CRI-O)
17:53:35      Loaded: loaded (/usr/local/lib/systemd/system/crio.service; enabled; vendor preset: disabled)
17:53:35      Active: active (running) since Tue 2021-12-14 20:53:30 UTC; 5s ago
17:53:35        Docs: https://github.com/cri-o/cri-o
17:53:35    Main PID: 119664 (crio)
17:53:35       Tasks: 12
17:53:35      Memory: 13.0M
17:53:35      CGroup: /system.slice/crio.service
17:53:35              └─119664 /usr/local/bin/crio
17:53:35 
17:53:35 Dec 14 20:53:29 fedoradocker32-443f80 crio[119664]: time="2021-12-14 20:53:29.871671086Z" level=info msg="No blockio config file specified, blockio not configured" file="blockio/blockio.go:37"
17:53:35 Dec 14 20:53:29 fedoradocker32-443f80 crio[119664]: time="2021-12-14 20:53:29.871708990Z" level=info msg="RDT not available in the host system" file="rdt/rdt.go:57"
17:53:35 Dec 14 20:53:29 fedoradocker32-443f80 crio[119664]: time="2021-12-14 20:53:29.872133728Z" level=info msg="Updated default CNI network name to " file="ocicni/ocicni.go:371"
17:53:35 Dec 14 20:53:29 fedoradocker32-443f80 crio[119664]: time="2021-12-14 20:53:29.872742783Z" level=debug msg="reading hooks from /usr/share/containers/oci/hooks.d" file="hooks/read.go:65"
17:53:35 Dec 14 20:53:29 fedoradocker32-443f80 crio[119664]: time="2021-12-14 20:53:29.878321889Z" level=debug msg="Golang's threads limit set to 114930" file="server/server.go:348"
17:53:35 Dec 14 20:53:29 fedoradocker32-443f80 crio[119664]: time="2021-12-14 20:53:29.992582345Z" level=debug msg="Sandboxes: [0xc00000c780 0xc00000c5a0 0xc00000c3c0 0xc00000c960]" file="server/server.go:496"
17:53:35 Dec 14 20:53:29 fedoradocker32-443f80 crio[119664]: time="2021-12-14 20:53:29.992717158Z" level=debug msg="Registered SIGHUP watcher for config" file="config/reload.go:32"
17:53:35 Dec 14 20:53:29 fedoradocker32-443f80 crio[119664]: time="2021-12-14 20:53:29.992829868Z" level=debug msg="Metrics are disabled" file="server/server.go:507"
17:53:35 Dec 14 20:53:30 fedoradocker32-443f80 crio[119664]: time="2021-12-14 20:53:30.013910578Z" level=debug msg="monitoring \"/usr/share/containers/oci/hooks.d\" for hooks" file="hooks/monitor.go:43"
17:53:35 Dec 14 20:53:30 fedoradocker32-443f80 systemd[1]: Started Container Runtime Interface for OCI (CRI-O).
17:53:35 [init.sh:291] INFO: Start Kubernetes
17:53:35 Init cluster using /var/run/crio/crio.sock
17:53:35 [init] Using Kubernetes version: v1.22.0
17:53:35 [preflight] Running pre-flight checks
17:53:35 error execution phase preflight: [preflight] Some fatal errors occurred:
17:53:35 	[ERROR Port-6443]: Port 6443 is in use
17:53:35 	[ERROR Port-10259]: Port 10259 is in use
17:53:35 	[ERROR Port-10257]: Port 10257 is in use
17:53:35 	[ERROR Port-2379]: Port 2379 is in use
17:53:35 	[ERROR Port-2380]: Port 2380 is in use
17:53:35 [preflight] If you know what you are doing, you can make a check non-fatal with `--ignore-preflight-errors=...`
17:53:35 To see the stack trace of this error execute with --v=5 or higher
17:53:35 [setup.sh:29] ERROR: bash ./init.sh
17:53:35 make: *** [Makefile:67: kubernetes-e2e] Error 1
17:53:35 [run.sh:80] ERROR: sudo -E PATH=/tmp/jenkins/workspace/kata-containers-2.0-tests-fedora-PR-crio-k8s-e2e-minimal/go/bin:/usr/local/go/bin:/usr/sbin:/sbin:/home/jenkins/.local/bin:/home/jenkins/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin bash -c make kubernetes-e2e
17:53:35 [jenkins_job_build.sh:215] ERROR: .ci/run.sh
```

There is a block of code that starts on [line 161](https://github.com/kata-containers/tests/blob/a1bd12f27288fee5112d4bf49a7e2ba3ba2f7075/integration/kubernetes/init.sh#L161) and spans until 171, that enable kubernetes debug, restart the kubelet service and then check it is up again. The failure happens on those checks for the kubelet service restart.

--- a/integration/kubernetes/init.sh
+++ b/integration/kubernetes/init.sh
@@ -169,6 +169,8 @@ start_kubernetes() {
 	local cri_socket_path="$2"
 	local cgroup_driver="$3"
 	local kubeadm_config_template="${SCRIPT_PATH}/kubeadm/config.yaml"
+	local kubelet_restart_wait="120"
+	local kubelet_restart_sleep="10"
 
 	info "Init cluster using ${cri_socket_path}"
 
@@ -200,16 +202,13 @@ start_kubernetes() {
 
 	# enable debug log for kubelet
 	sudo sed -i 's/.$/ --v=4"/' /var/lib/kubelet/kubeadm-flags.env
-	info "Kubelet options:"
+	info "Restart Kubelet with options:"
 	sudo cat /var/lib/kubelet/kubeadm-flags.env
 	sudo systemctl daemon-reload && sudo systemctl restart kubelet
 
-	for i in $(seq 4)
-	do
-		kubectl get nodes && break || sleep $i
-	done
-	kubectl get pods
-
+	info "Probing kubelet"
+	waitForProcess "$kubelet_restart_wait" "$kubelet_restart_sleep" \
+		"kubectl get nodes"
 }
 
 # Start the CRI runtime service.

